🎁 Use `commonDirectiveOptions` in all directives

agoose77 opened the pull request:
This PR tackles #1657. We will need an upstream `myst-theme` PR to ensure that these are actually set in the DOM.

I omitted the documentation directives `{myst:role}` et al. I didn't think they needed it.

agoose77 commented:
@fwkoch can you help clarify the changeset semver policy? I'm not clear for mystmd when we can make a minor bump.

agoose77 commented:
Requires https://github.com/jupyter-book/myst-theme/pull/500 to see some of the improvements.

fwkoch commented:
I put my thoughts about the semver policy here: https://github.com/jupyter-book/mystmd/pull/1658 - nothing official in that yet, just my opinions!

fwkoch reviewed (approved):
This is really nice for adding `class` to all node types, it adds that feature while simplifying the code.

My only hesitation is that it also adds `enumerated`/`enumerator` to all the node types. To me, this feels slightly less general than `class`, which is just a transparent string that themes may or may not support...

Enumeration gets picked up by the `enumerate` transforms and impacts the mdast processing. Testing a little bit, it seems ok... good actually... I can now enumerate many things that I could not before, and it _seems_ to just work. However, it makes the scope of this PR bigger than it might seem at first - it doesn't just add some simple properties to directives, it enriches (hopefully without breaking) fairly significant myst functionality.

Anyway, after going through all this, I'm still approving - I don't actually think there is a problem! Mostly just had to wrap my head around the implications of `enumerated`.

fwkoch commented on packages/myst-ext-exercise/src/exercise.ts:
```diff
@@ -36,19 +32,21 @@ export const exerciseDirective: DirectiveSpec = {
       enumerated: !nonumber,
       children: children as any[],
     };
+    addCommonDirectiveOptions(data, exercise);
```
Ah, I guess this is a small change requested, if you agree @agoose77

fwkoch replied:
Maybe this is just a documentation thing. `nonumber: true` is really just equivalent to `enumerated: false`, and I think it's fair that the former is just legacy support and the latter takes precedence?

Ugh - looking at the docs, there are pages where these directive properties are manually documented. e.g. https://mystmd.org/guide/exercises#exercise-directive - we should update those to use the `myst:directive` directive (though we would need to update that plugin to also import the `myst-ext-*` directives...).

agoose77 commented:
Yes @fwkoch, and helpful of you to explicitly surface that change --I made the same judgment that it feels like we should be able to identify and reference most (all) directives, but that isn't a small conceptual change by any means.

fwkoch commented on packages/myst-ext-exercise/src/exercise.ts:
```diff
@@ -8,12 +9,7 @@ export const exerciseDirective: DirectiveSpec = {
-    },
+    ...commonDirectiveOptions('exercise'),
     nonumber: {
       type: Boolean,
```
```suggestion
      type: Boolean,
      doc: 'Legacy flag to disable numbering of exercises; equivalent to `enumerated: false`',
```

fwkoch commented on packages/myst-ext-proof/src/proof.ts:
```diff
@@ -24,13 +24,7 @@ export const proofDirective: DirectiveSpec = {
-    },
+    ...commonDirectiveOptions('proof'),
     nonumber: {
       type: Boolean,
```
```suggestion
      type: Boolean,
      doc: 'Legacy flag to disable numbering of proofs; equivalent to `enumerated: false`',
```

fwkoch reviewed (commented):
Just thinking a little more about some documentation loose ends here...

agoose77 commented:
@fwkoch thanks for your review! Really good catch on the `enumerated` - `nonumber` overlap! I've updated the directives to prefer `nonumber` to `enumerated`, but respect the latter if given.

I've also reworked the docs to use `{myst}` or tabs for pairing syntax and rendered examples. **I removed reference to `sphinx-exercise`**, because at this point we are unifying things in a breaking manner, and we're trying to distance ourselves from the notion that MyST is 100% compatible with Sphinx.

fwkoch commented on docs/exercises.md:
```diff
@@ -113,21 +97,15 @@ factorial(4)
+The following options for solution directives are supported:
 
-  Removes the directive from the final output.
+:::{myst:directive} exercise
```
```suggestion
:::{myst:directive} solution
```

fwkoch commented:
The last thing I did was add `commonDirectiveOptions` to `solution` directive. I think that makes sense - you can now have, for example, `Solution 1` and `Solution 2` which may both be solutions to `Exercise 1`.
